File: src/audioWfc/wfc/hierarchy/NoteLevelNode.java
package audioWfc.wfc.hierarchy;

import audioWfc.wfc.HigherValues;
import audioWfc.musicTheory.OctavedNote;
import audioWfc.wfc.TileCanvas;
import audioWfc.wfc.CanvasAttributes;

import java.util.List;
import java.util.Random;

public class NoteLevelNode {
    private ChordLevelNode parent;
    private HigherValues higherValues;
    private CanvasAttributes<OctavedNote> noteAttributes;
    private TileCanvas<OctavedNote> canvas;

    public NoteLevelNode(ChordLevelNode parent, HigherValues higherValues, CanvasAttributes<OctavedNote> noteAttributes, Random random){
        this.parent = parent;
        this.noteAttributes = noteAttributes;
        this.canvas = new TileCanvas<>(
                noteAttributes.getSize(),
                noteAttributes.getOptions(),
                noteAttributes.getConstraints(),
                higherValues,
                random);
    }

    public List<OctavedNote> generate(){
        return canvas.generate();
    }

}

File: src/wfc/hierarchy/NoteLevelNode.ts
import { OctavedNote } from "../../music_theory/Note"
import { Random } from "../../util/Random"
import { HigherValues } from "../HigherValues"
import { TileCanvas, TileCanvasProps } from "../TileCanvas"
import { ConstraintHierarchy } from "../constraints/ConstraintInferrer/ConstraintHierarchy"
import { Constraint } from "../constraints/concepts/Constraint"

export class NoteLevelNode {
	private canvas: TileCanvas<OctavedNote>

	constructor(
		
		higherValues: HigherValues,
		random: Random,
		canvasProps?: TileCanvasProps<OctavedNote>,
		canvas?: TileCanvas<OctavedNote>
	) {
		if (!canvasProps && !canvas) {
			throw new Error("At least one of canvasProps or canvas must be provided.");
		}

		if (canvas) {
			this.canvas = canvas;
		} else {
			this.canvas = new TileCanvas<OctavedNote>(
				higherValues.melodyLength,
				canvasProps!,
				higherValues,
				random,
			);
		}
		
	}
	public inferConstraints(constraintHierarchy?: ConstraintHierarchy<OctavedNote>): Map<OctavedNote, Constraint<OctavedNote>[]>{
		if (constraintHierarchy != null){
			return constraintHierarchy.checkConstraintsGeneric(this.canvas.getConstraints(), this.canvas.getTiles())
		} else {
			const newConstraintHierarchy: ConstraintHierarchy<OctavedNote> = new ConstraintHierarchy<OctavedNote>(this.canvas.getHigherValues())
			return newConstraintHierarchy.checkConstraintsGeneric(this.canvas.getConstraints(), this.canvas.getTiles())
		}
	}
	
	public generate(): OctavedNote[] {
		return this.canvas.generate()
	}
}
